haystack_experimental/dataclasses/state.py

```diff
@@ -110,50 +127,29 @@ def set(
         self,
         key: str,
         value: Any,
-        handler_override: Optional[Callable[[Any, Any], Any]] = None,
-        force: bool = False
+        handler_override: Optional[Callable[[Any, Any], Any]] = None
     ) -> None:
         """
         Set or merge a value in the state according to schema rules.
 
         Value is merged or overwritten according to these rules:
-          - If force=True, just overwrite
+          - if the current value in the state is None, replace it
           - else if handler_override is given, use that
-          - else if the schema for 'key' has a custom handler, use it
-          - else use default merge logic
+          - else use the handler defined in the schema for 'key'
 
         :param key: Key to store the value under
         :param value: Value to store or merge
-        :param handler_override: Optional function to override default merge behavior
-        :param force: If True, overwrites existing value without merging
+        :param handler_override: Optional function to override the default merge behavior
         """
-        # If key not in schema, we consider no special merging => just store
-        definition = self.schema.get(key, {})
-        declared_type = definition.get("type")
-        declared_handler = definition.get("handler")
-
-        if force or (not declared_type and not declared_handler and not handler_override):
-            self._data[key] = value
-            return
+        # If key not in schema, we throw an error
+        definition = self.schema.get(key, None)
+        if definition is None:
+            raise ValueError(f"State: Key '{key}' not found in schema. Schema: {self.schema}")
 
+        # if the current value is None just replace
```

Comment:
Is this comment accurate? Aren't we getting the current value from state?

Reply:
good catch! this was a leftover